Create a Docker image for use as a sidecar log collector that sends logs to Google Cloud Logging

a-robinson opened the pull request:
If this looks good, I'll make an almost identical elasticsearch version. Mostly closes #7006.

@satnam6502

a-robinson committed:
Create a Docker image for use as a sidecar log collector that sends logs
to Google Cloud Logging.

a-robinson commented:
ping

satnam6502 commented on contrib/logging/fluentd-sidecar-gcp/Dockerfile:
```diff
@@ -0,0 +1,35 @@
+# Run the config generator to get the config files in place and start Fluentd.
+# We have to run the config generator at runtime rather than now so that it can
+# incorporate the files provided in the environment variable in its config.
+CMD /usr/local/sbin/config_generator.sh && /usr/sbin/google-fluentd --log /var/log/google-fluentd/google-fluentd.log --use-v1-config --suppress-repeated-stacktrace
```
What about `-qq` to keep the Fluentd log file under control? Otherwise setting log rotation for it might be a plan.

a-robinson replied:
Good point. I was initially going to just let it write its logs to stdout for the node-level fluentd to collect, but decided that those logs will probably just confuse most users, so I changed it to write to a file but forgot to re-add the -qq.

a-robinson replied:
That also makes it better to do redirection rather than using the --log flag, to overwrite the gunk generated by the agent starting when it's first installed.

mulloymorrow replied:
Have you considered leaving the logging-level-specification to the system directive rather than as a run option in the dockerfile? 

```
<system>
  # equal to -qq option
  log_level error
  # equal to --without-source option
  without_source
  # ...
</system>
```

satnam6502 commented on contrib/logging/fluentd-sidecar-gcp/README.md:
```diff
@@ -0,0 +1,22 @@
+This shouldn't be necessary if your container writes its logs to stdout or stderr, since the Kubernetes cluster's default logging infrastructure will collect that automatically, but this is useful if your application logs to a specific file in its filesystem and can't easily be changed.
+
+In order to make this work, you have to add a few things to your pod config:
+1. A second container, using the `gcr.io/google_containers/fluentd-sidecar-gcp:1.0` image to send the logs to Google Cloud Logging.
```
Please can this be formatted as a list with each item starting on a new line? Thanks.

a-robinson replied:
Gah, I thought that this would make it a list. Markdown fail :(

satnam6502 commented on contrib/logging/fluentd-sidecar-gcp/config_generator.sh:
```diff
@@ -0,0 +1,37 @@
+  cat > "/etc/google-fluentd/files/${filename}" << EndOfMessage
+<source>
+  type tail
+  format none
```
It's a shame that if any files are in JSON format then any extra useful information won't get propagated. But for this I think this is fine.

a-robinson replied:
Yeah, in order for this to be general, we can't assume that logs will be in JSON. And this model of collecting logs is most useful for older applications that can't be modified to write to stdout, and I'd expect fewer older applications to write JSON logs.

satnam6502 commented:
Looks great, thanks, just a few minor nits.

a-robinson committed:
Add an example usage of the cloud logging sidecar image to the same directory
as the image components. It doesn't really belong under examples/ since
it's so specific to GCE/GKE.

a-robinson commented:
Pushed new version with the improvements.

mulloymorrow commented:
+1 elasticsearch version
